### ISSUE TITLE
[React 19] infinite loop when using React lazy & functional formAction together.

### ISSUE THREAD

**Issue by Henry-Hong** [facebook/react#29930]
### Version
"react": "19.0.0-rc-107a2f8c3e-20240617",
"react-dom": "19.0.0-rc-107a2f8c3e-20240617"
OS : windows 11
Chrome : 125.0.6422.142

### Overview
We have a great feature 'functional form action' in react 19.
`<form action={functionalFormAction}`
<br>

but with React lazy, it doesn't seem to work properly.
It causes infinite loop depends on its use cases.

![image](https://github.com/facebook/react/assets/17701725/bd5e67a6-e990-45b8-a57a-531e81617863)

<br>

1. case1 : infinite loop (React lazy without `<Suspense/>`. All dynamic codes are inside form component)
2. case2 : no infinite loop (React lazy with `<Suspense/>`. All dynamic codes are inside form component)
3. case3 : infinite loop (React lazy with `<Suspsne />` & has dependency on useFormStatus. All dynamic code are inside form component)

### Related Issues
#29235 Bug: Nested lazy components cause rerendering
#27573 Bug: Nested components imported with React.lazy cause rerendering

### Reproduction
Go to [CodeSandBox](https://codesandbox.io/p/devbox/devbox-test-hhr-zm6z3k?file=%2Fsrc%2FApp.tsx%3A11%2C46)
Try Case2 at first, (submit form by clicking button)
Try Case3 at second, (it causes infinite loop so it makes js blocked)
Try Case1 at the last (it causes infinite loop so it makes js blocked)

### Preview
App.tsx
![image](https://github.com/facebook/react/assets/17701725/1300ae79-1c3c-4f79-bd82-06e723ed6b9f)

Case1.tsx
![image](https://github.com/facebook/react/assets/17701725/761d08d4-8fd7-4692-b0a8-954f1124df48)

Case 1 Result
![image](https://github.com/facebook/react/assets/17701725/f18f610c-7b4b-42fe-88b9-4bd8addc5d86)

Case 2 Result
![image](https://github.com/facebook/react/assets/17701725/17b3eba5-ae81-4017-83de-bb3a17e5b4e3)

Case 3 Result
![image](https://github.com/facebook/react/assets/17701725/4eac93e2-e498-483c-9a8c-7bc99c2b20b9)

**Comment by eps1lon** [comment]
You are creating a new, lazy component within render. The component type is therefore changing on every render. Then React has to re-render when the lazy resolves but encounters a new lazy component since you called `lazy()` again. This is expected behavior.

Do you still enter an infinite loop if you move the component creation outside of render?

```diff
+const TargetComponent = lazy(...)

 function Case() {
-  const TargetComponent = lazy(...)
   return <TargetComponent />
 }
```

I'd recommend splitting these up into different repros. It's hard to follow how you use the repro and what you're actually reproducing.

**Comment by Henry-Hong** [comment]
> Do you still enter an infinite loop if you move the component creation outside of render?

I was able to prevent infinite loop for 'Case 1'. But unfortunately, 'Case 3' is still suffering from infinite loop.

And, moving the `React.lazy` outside of render has limits. Due to the inability to utilize component props, 'React.lazy' can only be used statically.

for example, I have an <Svg /> component that takes file names as props and renders dynamically based on them. According to the proposed solution, this code can no longer be used.

![image](https://github.com/facebook/react/assets/17701725/e61fa327-50b3-4779-9a5d-7f6005dc5958)

**Comment by eps1lon** [comment]
> And, moving the React.lazy outside of render has limits. Due to the inability to utilize component props, 'React.lazy' can only be used statically.

You need to implement a client cache then. React doesn't have that capabiltiy at the moment. Unconditionally creating `React.lazy` during render is not supported. Just like you couldn't create a new Promise during render and pass it to `use`.

It's probably better to use some form of codegen to create the icons anyway so that all available icons are known statically.

> I was able to prevent infinite loop for 'Case 1'. But unfortunately, 'Case 3' is still suffering from infinite loop.

Do you have a repro just for that case with the `React.lazy` created outside of render?

**Comment by Henry-Hong** [comment]
Thank you for replying.

> It's probably better to use some form of codegen to create the icons anyway so that all available icons are known statically.

I thought the same thing, and I checked that lots of icon libraries are supporting icons in the form of codegen which is statically generated. (like [lucide](https://github.com/lucide-icons/lucide/blob/main/tools/build-icons/cli.mjs))

> Do you have a repro just for that case with the React.lazy created outside of render?

I was trying to repro the infinite rerender on my local machine, but it doesn't make any infinite loop when React.lazy created outside of render. 

I think it's environment issue related with Codesandbox which is not that import at this moment. 

Thank you for supporting, this conversation helped me a lot understanding react deeper.
